### DISCUSSION TITLE
Support for self-adding a new language kernel and cell?

### DISCUSSION THREAD

**Discussion by bmitc** [dotnet/interactive#2255] [Q&A]
Hello. I am working for a company that has our [own custom language](https://github.com/sifive/wake). It is an ML-like but is not a .NET language. Being familiar with .NET Interactive notebooks (mainly through F#), I am wondering if it is possible to self-add a custom language kernel to .NET Interactive such that users could pull in a package or VS Code extension or something and use this custom language in .NET Interactive notebooks. This is sort of treating .NET Interactive as a notebook platform.

The benefits are that:

1. We would get all of the polyglot and markdown support of .NET Interactive notebooks for free
2. We would get Visual Studio Code integration for free, which is a plus since we already use VS Code as the primary IDE for this custom language and even have an [extension for it](https://marketplace.visualstudio.com/items?itemName=sifive.wake).
3. This would enable streamlined internal learning of the language through the notebook paradigm.

If possible, it would be great if it's possible to also add the language to the cell selection window.

**Comment by jonsequitur** [comment]
It's fairly straightforward to add languages. This question came up on Twitter recently, in fact: https://twitter.com/GamerBlom/status/1563154306945417216

Here's another example: https://twitter.com/bvfors/status/1436465058528337937

These show two slightly different approaches. The first one has a .NET wrapper whereas the second one communicates purely via messages over standard I/O.
